### PR TITLE
Added pytorch frontend block_diag and its tests

### LINKED ISSUE #19198: block_diag

### PR DESCRIPTION
**Pull request by vishwajit-vishnu**
Close #19198

### CODE REVIEW

**Comment by vishwajit-vishnu** [comment]
1 backend test is failing due to already faulty paddle's scatter_nd function

**Comment by vishwajit-vishnu** [comment]
@Dsantra92 Can you please review and merge

**Comment by vishwajit-vishnu** [comment]
@MuhammedAshraf2020 can you please review and merge

**Comment by vishwajit-vishnu** [comment]
@ivy-leaves @MuhammedAshraf2020 can you please review or add any suggestions for the PR

**Comment by vishwajit-vishnu** [comment]
@ivy-seed @MuhammedAshraf2020  can you please review this PR??

**Comment by vishwajit-vishnu** [comment]
@ivy-leaves @ivy-seed @MuhammedAshraf2020 Can you please review it ?? Its been more than two weeks now. Atleast looking for some comment or some suggestions

**Comment by MuhammedAshraf2020** [comment]
Hey @vishwajit-vishnu ,
Apologies for the latency. Thank you for your valuable contribution. Your PR will be reviewed very soon.

**Comment by vishwajit-vishnu** [comment]
@AnnaTz Can you please review it

**Review by AnnaTz — CHANGES_REQUESTED** [review]

File: ivy/functional/frontends/torch/miscellaneous_ops.py [hunk 1]
```diff
@@ -413,6 +414,54 @@ def cov(input, /, *, correction=1, fweights=None, aweights=None):
...
+def block_diag(*tensors):
+    shapes_list = [ivy.shape(t) for t in tensors]
+    # TODO: Add ivy function to return promoted dtype for multiple tensors at once
+    promoted_dtype = ivy.as_ivy_dtype(tensors[0].dtype)
+    for idx in range(1, len(tensors)):
+        promoted_dtype = torch_frontend.promote_types_torch(
+           tensors[idx - 1].dtype, tensors[idx].dtype
+        )
+
+    inp_tensors = [ivy.asarray(t, dtype=promoted_dtype) for t in tensors]
```
[the remark this answers is not in the document]
Comment (vishwajit-vishnu) [reply ivy/functional/frontends/torch/miscellaneous_ops.py:430]: @AnnaTz  Its not redundant as we are copying the array to a specific location in zeros array and not promoting their dtype while copying. So Its failing the tests if we do not do this step

**Comment by vishwajit-vishnu** [comment]
@AnnaTz  thanks for review. I have made the changes. You can merge it. Also wanted to ask if this can be implemented as a mixed function where for torch and scipy the backend will return the value but for other backends it will raise a NotImplementedError and we can use the current code  in a ivy function to give functionality for other backends

**Comment by AnnaTz** [comment]
> @AnnaTz thanks for review. I have made the changes. You can merge it. Also wanted to ask if this can be implemented as a mixed function where for torch and scipy the backend will return the value but for other backends it will raise a NotImplementedError and we can use the current code in a ivy function to give functionality for other backends

There is one more line error to resolve.
Yes we could do that. In mixed function we don't get NotImplementedError error though. For the frameworks that don't have a corresponding function we use a compositional implementation which is the main ivy function made of other ivy functions, in this case the code you wrote for the frontend would work.

**Review by AnnaTz — APPROVED** [review]
Great, thanks!

**Comment by AnnaTz** [comment]
> > Yes we could do that. In mixed function we don't get NotImplementedError errors though. For the frameworks that don't have a corresponding function we use a compositional implementation which is the main ivy function made of other ivy functions, in this case the code you wrote for the frontend would work.
> 
> You can merge the changes Can you create an issue to create ivy function for block_diag and I can add ivy functionn for that. I read documentation here [mixed function](https://unify.ai/docs/ivy/overview/deep_dive/function_types.html#mixed-functions), but compositional functions are used only when all backends do not have this function block_diag. But torch and scipy has that. I did not find any example on mixed function. Sorry if I misunderstood and again thanks for the help

I added `block_diag` [here](https://github.com/unifyai/ivy/issues/3856). Feel free to open an issue and start working on that!

**Comment by AnnaTz** [comment]
> > Yes we could do that. In mixed function we don't get NotImplementedError errors though. For the frameworks that don't have a corresponding function we use a compositional implementation which is the main ivy function made of other ivy functions, in this case the code you wrote for the frontend would work.
> 
> You can merge the changes Can you create an issue to create ivy function for block_diag and I can add ivy functionn for that. I read documentation here [mixed function](https://unify.ai/docs/ivy/overview/deep_dive/function_types.html#mixed-functions), but compositional functions are used only when all backends do not have this function block_diag. But torch and scipy has that. I did not find any example on mixed function. Sorry if I misunderstood and again thanks for the help

image

**Pull request merged**